src/graphene.cpp

```diff
@@ -0,0 +1,1471 @@
+            pfrom->PushMessage(NetMsgType::BLOCK, *pblock);
+            LOG(GRAPHENE,
+                "Sent regular block instead - encountered error when creating graphene block for peer %s: %s\n",
+                pfrom->GetLogName(), e.what());
+        }
+    }
+    else
+    {
+        dosMan.Misbehaving(pfrom, 100);
+
+        return;
+    }
+
+    pfrom->blocksSent += 1;
+}
+
+bool IsGrapheneBlockValid(CNode *pfrom, const CBlockHeader &header)
+{
+    // check block header
+    CValidationState state;
+    if (!CheckBlockHeader(header, state, true))
+    {
+        return error("Received invalid header for graphene block %s from peer %s", header.GetHash().ToString(),
+            pfrom->GetLogName());
+    }
+    if (state.Invalid())
+    {
+        return error("Received invalid header for graphene block %s from peer %s", header.GetHash().ToString(),
+            pfrom->GetLogName());
+    }
+
+    return true;
+}
+
+bool HandleGrapheneBlockRequest(CDataStream &vRecv, CNode *pfrom, const CChainParams &chainparams)
+{
+    if (!pfrom->GrapheneCapable())
+    {
+        dosMan.Misbehaving(pfrom, 100);
+        return error("Graphene block message received from a non graphene block node, peer=%d", pfrom->GetId());
+    }
+
+    // Check for Misbehaving and DOS
+    // If they make more than 20 requests in 10 minutes then disconnect them
+    {
+        LOCK(cs_vNodes);
+        if (pfrom->nGetGrapheneLastTime <= 0)
+            pfrom->nGetGrapheneLastTime = GetTime();
+        uint64_t nNow = GetTime();
+        pfrom->nGetGrapheneCount *= std::pow(1.0 - 1.0 / 600.0, (double)(nNow - pfrom->nGetGrapheneLastTime));
+        pfrom->nGetGrapheneLastTime = nNow;
+        pfrom->nGetGrapheneCount += 1;
+        LOG(GRAPHENE, "nGetGrapheneCount is %f\n", pfrom->nGetGrapheneCount);
+        if (chainparams.NetworkIDString() == "main") // other networks have variable mining rates
+        {
+            if (pfrom->nGetGrapheneCount >= 20)
+            {
+                dosMan.Misbehaving(pfrom, 100); // If they exceed the limit then disconnect them
+                return error("sending too many GET_GRAPHENE messages");
+            }
+        }
+    }
+
+    CMemPoolInfo mempoolinfo;
+    CInv inv;
+    vRecv >> inv >> mempoolinfo;
+    graphenedata.UpdateInBoundMemPoolInfo(::GetSerializeSize(mempoolinfo, SER_NETWORK, PROTOCOL_VERSION));
+
+    // Message consistency checking
+    if (!(inv.type == MSG_GRAPHENEBLOCK) || inv.hash.IsNull())
+    {
+        dosMan.Misbehaving(pfrom, 100);
+        return error("invalid GET_GRAPHENE message type=%u hash=%s", inv.type, inv.hash.ToString());
+    }
+
+    CBlock block;
+    {
+        LOCK(cs_main);
+        BlockMap::iterator mi = mapBlockIndex.find(inv.hash);
+        if (mi == mapBlockIndex.end())
+        {
+            dosMan.Misbehaving(pfrom, 100);
+            return error(
+                "Peer %s (%d) requested nonexistent block %s", pfrom->addrName.c_str(), pfrom->id, inv.hash.ToString());
```

Comment:
What are the repercussions of sending an error message back or just silently dropping?  I do not think this results in a DoS attack any worse than all the other ways you can send random stuff to our client (like lots of fake INVs).
But an advantage of allowing it is to enable unsolicited graphene requests.  If a node becomes aware of a block through any means, it could request it speculatively from its lowest latency peer.  This would likely especially help GFC penetration -- the block announcement (header) could come from China, but your node could request it from a node outside the GFC.

Reply:
TBH I copied the basic logic from XThin, so I'm not strongly opinionated about how to handle this case. However, my take on this condition is that it's checking if a block was requested for which we have no knowledge. Whereas it sounds like you're thinking of a scenario (and I might be totally misunderstanding your point here) where we do have the block, but we just never announced it to the requesting peer.

Reply:
That is the scenario I'm thinking of.  The problem is that a peer will not risk getting banned.  So it won't make speculative requests unless there's no penalty for having data its peer does not have.

Reply:
Ok, I see. So I'm just going to drop the `dosMan.Misbehaving(pfrom, 100)` from that case then.